forte/data/extractor/base_extractor.py

```diff
@@ -0,0 +1,205 @@
+        return obj
+
+    @property
+    def entry_type(self) -> Type[Annotation]:
+        return self.config.entry_type
+
+    @property
+    def vocab_method(self) -> str:
+        return self.config.vocab_method
+
+    def get_pad_value(self) -> Union[None, int, List[int]]:
+        if self.vocab:
+            return self.vocab.get_pad_value()
+        else:
+            return None
+
+    def items(self) -> Iterable[Tuple[Hashable, int]]:
+        assert self.vocab
+        return self.vocab.items()
+
+    def size(self) -> int:
+        assert self.vocab
+        return len(self.vocab)
+
+    def add(self, element: Hashable):
+        assert self.vocab
+        return self.vocab.add_element(element)
+
+    def has_element(self, element: Hashable) -> bool:
+        assert self.vocab
+        return self.vocab.has_element(element)
+
+    def element2repr(self, element: Hashable) -> Union[int, List[int]]:
+        assert self.vocab
+        return self.vocab.element2repr(element)
+
+    def id2element(self, idx: int) -> Any:
+        assert self.vocab
+        return self.vocab.id2element(idx)
+
+    def get_dict(self) -> Dict[Hashable, int]:
+        assert self.vocab
+        return self.vocab.get_dict()
+
+    def predefined_vocab(self, predefined: Union[Set, List]):
+        """This function will add elements from the
+        passed-in predefined set to the vocab. Different
+        extractors might have different strategies to add
+        these elements. Override this function if necessary.
+
+        Args:
+            predefined (Union[Set, List]): A set or list contain
+                elements to be added into the vocabulary.
+        """
+        for element in predefined:
+            self.add(element)
+
+    def update_vocab(self, pack: DataPack,
+                    instance: Annotation):
+        """This function will extract the feature from
+        instance and add element in the feature to vocabulary.
+
+        Args:
+            pack (Datapack): The datapack that contains the current
+                instance.
+            instance (Annotation): The instance from which the
+                extractor will extractor feature.
+        """
+        raise NotImplementedError()
+
+    def extract(self, pack: DataPack,
+                instance: Annotation) -> Feature:
+        """This function will extract feature from
+        one instance in the pack.
+
+        Args:
+            pack (Datapack): The datapack that contains the current
+                instance.
+            instance (Annotation): The instance from which the
+                extractor will extractor feature.
+
+        Returns:
+            Feature: a feature that contains the extracted data.
+        """
+        raise NotImplementedError()
+
+    def add_to_pack(self, pack: DataPack, instance: Annotation,
+                    prediction: Any):
+        """This function will remove the original entry and
+        add prediction to the pack.
```

Comment:
Please explain what user would implement here.

By the way, what do you mean by "remove the original entry"?

Reply:
Got it.
The `remove` pack is a tricky thing. It is for the evaluation purpose.
So if the pack originally have "those features" in the pack, what should we do when we try to add the prediction to the pack?
For example, in the evaluation during the training process. The golden value is already there. The train_preprocessor will hold a pack as reference and copy a pack to do prediction. For the pack that is copied and used as prediction, I need to first remove the original data there and then add the predicted value.
Do you have any suggest on handling possible confusion here?

Reply:
I actually think we should add a separate processor just for the removal part. The extractor can assume that the input is clean.

Think of this scenario, the user may provide plain text for the system to extract, then we don't have the removal part.